File: LevelTransition.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionListener;

/**
 * Handles level transitions in the game.
 */
public class LevelTransition {
    private boolean isTransitioning = false;
    private Level currentLevel;
    private Level nextLevel;
    private float transitionAlpha = 0.0f;
    private Timer transitionTimer;
    private static final float TRANSITION_STEP = 0.05f;
    private boolean fadeIn = false;
    
    /**
     * Constructor for the level transition.
     */
    public LevelTransition() {
        this.transitionTimer = null;
    }
    
    /**
     * Start a level transition.
     * 
     * @param currentLevel The level transitioning from
     * @param nextLevel The level transitioning to
     * @param finishedCallback The callback to run when the transition is complete
     */
    public void startTransition(Level currentLevel, Level nextLevel, Runnable finishedCallback) {
        this.currentLevel = currentLevel;
        this.nextLevel = nextLevel;
        this.isTransitioning = true;
        this.transitionAlpha = 0.0f;
        this.fadeIn = false;
        
        if (transitionTimer != null && transitionTimer.isRunning()) {
            transitionTimer.stop();
        }
        
        ActionListener transitionAction = e -> {
            if (!fadeIn) {
                // Fading out
                transitionAlpha += TRANSITION_STEP;
                if (transitionAlpha >= 1.0f) {
                    transitionAlpha = 1.0f;
                    fadeIn = true;
                }
            } else {
                // Fading in
                transitionAlpha -= TRANSITION_STEP;
                if (transitionAlpha <= 0.0f) {
                    transitionAlpha = 0.0f;
                    isTransitioning = false;
                    ((Timer)e.getSource()).stop();
                    finishedCallback.run();
                }
            }
        };
        
        transitionTimer = new Timer(50, transitionAction);
        transitionTimer.start();
    }
    
    /**
     * Draw the transition effect.
     * 
     * @param g The graphics context
     * @param width The width of the game area
     * @param height The height of the game area
     */
    public void draw(Graphics g, int width, int height) {
        if (!isTransitioning) return;
        
        if (fadeIn) {
            // When fading in, draw the next level text
            drawLevelText(g, width, height, nextLevel);
        } else {
            // When fading out, draw the current level text
            drawLevelText(g, width, height, currentLevel);
        }
        
        // Draw overlay with current alpha
        g.setColor(new Color(0, 0, 0, (int)(transitionAlpha * 255)));
        g.fillRect(0, 0, width, height);
    }
    
    private void drawLevelText(Graphics g, int width, int height, Level level) {
        g.setColor(new Color(
            GameColors.ACCENT_YELLOW.getRed(),
            GameColors.ACCENT_YELLOW.getGreen(),
            GameColors.ACCENT_YELLOW.getBlue(),
            (int)((1.0f - Math.abs(transitionAlpha - 0.5f) * 2) * 255)
        ));
        
        Font titleFont = new Font("Monospaced", Font.BOLD, 36);
        g.setFont(titleFont);
        
        String levelText = "LEVEL " + level.getLevelNumber();
        FontMetrics fm = g.getFontMetrics();
        
        int textX = (width - fm.stringWidth(levelText)) / 2;
        int textY = height / 2 - 20;
        
        g.drawString(levelText, textX, textY);
        
        // Draw level description
        Font descFont = new Font("Monospaced", Font.PLAIN, 16);
        g.setFont(descFont);
        
        String[] descriptions = {
            "MAX HEALTH: " + level.getPlayerMaxHealth(),
            "ENEMY HEALTH: +" + (int)((level.getEnemyHealthMultiplier() - 1.0f) * 100) + "%",
            "ENEMY DAMAGE: +" + (int)((level.getEnemyDamageMultiplier() - 1.0f) * 100) + "%"
        };
        
        fm = g.getFontMetrics();
        textY += 40;
        
        for (String desc : descriptions) {
            textX = (width - fm.stringWidth(desc)) / 2;
            g.drawString(desc, textX, textY);
            textY += 30;
        }
        
        String readyText = "GET READY!";
        textX = (width - fm.stringWidth(readyText)) / 2;
        textY += 20;
        g.drawString(readyText, textX, textY);
    }
    
    /**
     * Check if a transition is in progress.
     * 
     * @return true if transitioning, false otherwise
     */
    public boolean isTransitioning() {
        return isTransitioning;
    }
    
    /**
     * Stop the current transition.
     */
    public void stopTransition() {
        if (transitionTimer != null && transitionTimer.isRunning()) {
            transitionTimer.stop();
        }
        isTransitioning = false;
    }
}

File: levelTransition.js
/**
 * Handles level transitions in the game.
 */
class LevelTransition {
    constructor() {
        this.isTransitioning = false;
        this.currentLevel = null;
        this.nextLevel = null;
        this.transitionAlpha = 0.0;
        this.fadeIn = false;
        this.transitionStep = 0.05;
        this.finishedCallback = null;
        this.lastTransitionTime = 0;
    }
    
    /**
     * Start a level transition.
     * 
     * @param {Level} currentLevel The level transitioning from
     * @param {Level} nextLevel The level transitioning to
     * @param {Function} finishedCallback The callback to run when the transition is complete
     */
    startTransition(currentLevel, nextLevel, finishedCallback) {
        this.currentLevel = currentLevel;
        this.nextLevel = nextLevel;
        this.isTransitioning = true;
        this.transitionAlpha = 0.0;
        this.fadeIn = false;
        this.finishedCallback = finishedCallback;
        this.lastTransitionTime = 0;
    }
    
    /**
     * Update the transition effect.
     * 
     * @param {number} timestamp Current timestamp for animations
     */
    update(timestamp) {
        if (!this.isTransitioning) return;
        
        // Update every 50ms (similar to the Java Timer's delay)
        if (timestamp - this.lastTransitionTime < 50) return;
        this.lastTransitionTime = timestamp;
        
        if (!this.fadeIn) {
            // Fading out
            this.transitionAlpha += this.transitionStep;
            if (this.transitionAlpha >= 1.0) {
                this.transitionAlpha = 1.0;
                this.fadeIn = true;
            }
        } else {
            // Fading in
            this.transitionAlpha -= this.transitionStep;
            if (this.transitionAlpha <= 0.0) {
                this.transitionAlpha = 0.0;
                this.isTransitioning = false;
                if (this.finishedCallback) {
                    this.finishedCallback();
                }
            }
        }
    }
    
    /**
     * Draw the transition effect.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} width The width of the game area
     * @param {number} height The height of the game area
     */
    draw(ctx, width, height) {
        if (!this.isTransitioning) return;
        
        if (this.fadeIn) {
            // When fading in, draw the next level text
            this.drawLevelText(ctx, width, height, this.nextLevel);
        } else {
            // When fading out, draw the current level text
            this.drawLevelText(ctx, width, height, this.currentLevel);
        }
        
        // Draw overlay with current alpha
        ctx.fillStyle = `rgba(0, 0, 0, ${this.transitionAlpha})`;
        ctx.fillRect(0, 0, width, height);
    }
    
    /**
     * Draw the level text.
     * 
     * @param {CanvasRenderingContext2D} ctx The canvas context
     * @param {number} width The width of the game area
     * @param {number} height The height of the game area
     * @param {Level} level The level to display
     */
    drawLevelText(ctx, width, height, level) {
        // Calculate text opacity based on transition progress
        let textAlpha = 1.0 - Math.abs(this.transitionAlpha - 0.5) * 2;
        ctx.fillStyle = this.hexToRgba(GameColors.ACCENT_YELLOW, textAlpha);
        
        ctx.font = 'bold 36px monospace';
        let levelText = 'LEVEL ' + level.levelNumber;
        let textWidth = ctx.measureText(levelText).width;
        ctx.fillText(levelText, (width - textWidth) / 2, height / 2 - 20);
        
        // Draw level description
        ctx.font = '16px monospace';
        
        let descriptions = [
            'MAX HEALTH: ' + level.playerMaxHealth,
            'ENEMY HEALTH: +' + Math.floor((level.enemyHealthMultiplier - 1.0) * 100) + '%',
            'ENEMY DAMAGE: +' + Math.floor((level.enemyDamageMultiplier - 1.0) * 100) + '%'
        ];
        
        let y = height / 2 + 20;
        for (let desc of descriptions) {
            textWidth = ctx.measureText(desc).width;
            ctx.fillText(desc, (width - textWidth) / 2, y);
            y += 30;
        }
        
        let readyText = 'GET READY!';
        textWidth = ctx.measureText(readyText).width;
        ctx.fillText(readyText, (width - textWidth) / 2, y + 20);
    }
    
    /**
     * Convert hex color to rgba.
     * 
     * @param {string} hex The hex color
     * @param {number} alpha The alpha value
     * @return {string} The rgba color string
     */
    hexToRgba(hex, alpha) {
        let r = parseInt(hex.slice(1, 3), 16);
        let g = parseInt(hex.slice(3, 5), 16);
        let b = parseInt(hex.slice(5, 7), 16);
        return `rgba(${r}, ${g}, ${b}, ${alpha})`;
    }
    
    /**
     * Stop the current transition.
     */
    stopTransition() {
        this.isTransitioning = false;
    }
}